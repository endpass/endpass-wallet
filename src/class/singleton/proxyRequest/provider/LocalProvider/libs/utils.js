import regPath from 'path-to-regexp';

const utils = {
  getRouteKey(routesRegexp, routePath) {
    return Object.keys(routesRegexp).find(key => {
      const re = routesRegexp[key];
      return re.test(routePath);
    });
  },
  getPath(serverUrl, url) {
    const levels = url.split(serverUrl);
    const keysPath = levels.pop() || '';
    const ret = keysPath[0] === '/' ? keysPath.substring(1) : keysPath;

    return ret;
  },
  getPathArgs(routesRegexp, key, routePath) {
    const keys = regPath.parse(key);
    const re = routesRegexp[key];
    re.lastIndex = 0; // drop exec position

    const values = re.exec(routePath);
    return keys.reduce((map, item, index) => {
      if (item && item.name) {
        map[item.name] = values[index];
      }
      return map;
    }, {});
  },
};

export default utils;
