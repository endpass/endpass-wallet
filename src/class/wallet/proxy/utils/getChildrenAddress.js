export default function(hdWallet, offset, limit) {
  return [...Array(limit)].map((_, i) =>
    hdWallet
      .deriveChild(offset + i)
      .getWallet()
      .getChecksumAddressString(),
  );
}
