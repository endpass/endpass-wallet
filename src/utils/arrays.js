export const uniq = arr =>
  arr.reduce((acc, item) => (acc.includes(item) ? acc : acc.concat(item)), []);

export const mapArrayByProp = (arr, prop) =>
  arr.reduce((acc, item) => {
    const target = item[prop];

    if (target) {
      return Object.assign(acc, {
        [target]: item,
      });
    }

    return acc;
  }, {});

export default {
  uniq,
  mapArrayByProp,
};
