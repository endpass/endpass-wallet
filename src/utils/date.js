import dayjs from 'dayjs';
import relativeTime from 'dayjs/plugin/relativeTime';

/**
 * Setting up dayjs globally
 */
dayjs.extend(relativeTime);

export const formateDate = date => dayjs(date).format('YYYY-MM-DD H:mm');

export const fromNow = date => dayjs(date).fromNow();

export default {
  formateDate,
  fromNow,
};
