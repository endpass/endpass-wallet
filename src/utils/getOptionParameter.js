export default function getOptionParameter(item, value) {
  return item instanceof Object ? item[value] : item;
}
