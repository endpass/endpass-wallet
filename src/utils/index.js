import store from '../store';

export { default as http } from './http';

export const kebabToCamel = input =>
  input.replace(/-([a-z])/g, g => g[1].toUpperCase());

export const getInitializedValueFromStore = (module, field) => {
  return new Promise(resolve => {
    const value = module[field];

    if (value === null) {
      const unwatch = store.watch(
        () => module[field],
        newValue => {
          unwatch();
          resolve(newValue);
        },
      );
    } else {
      resolve(value);
    }
  });
};
