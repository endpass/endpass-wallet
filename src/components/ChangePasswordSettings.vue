<template>
  <v-form
    :is-form-valid="isFormValid"
    @submit="handleFormSubmit"
  >
    <label class="label">Change Password</label>
    <v-password
      v-model="oldPassword"
      v-validate="'required|min:8'"
      :disabled="isLoading"
      :error="errors.first('oldPassword')"
      name="oldPassword"
      data-vv-name="oldPassword"
      data-vv-as="Old Password"
      placeholder="Old Password"
      data-test="input-old-password"
    />
    <v-password
      v-model="newPassword"
      v-validate="'required|min:8'"
      :disabled="isLoading"
      :error="errors.first('newPassword')"
      name="newPassword"
      data-vv-name="newPassword"
      data-vv-as="New Password"
      placeholder="New Password"
      data-test="input-new-password"
    />
    <v-button
      :loading="isLoading"
      :disabled="!isFormValid"
      class-name="is-primary is-medium"
      data-test="submit-change-password"
    >
      Change Password
    </v-button>
  </v-form>
</template>

<script>
import { mapActions } from 'vuex';
import { matchString } from '@endpass/utils/strings';
import formMixin from '@/mixins/form';

export default {
  name: 'ChangePasswordSettings',

  data: () => ({
    isLoading: false,
    oldPassword: null,
    newPassword: null,
  }),

  methods: {
    ...mapActions('accounts', [
      'updateWallets',
      'updateWalletsWithNewPassword',
    ]),

    async handleFormSubmit() {
      try {
        this.isLoading = true;

        const res = await this.updateWalletsWithNewPassword({
          password: this.oldPassword,
          newPassword: this.newPassword,
        });

        if (!res) {
          this.handleSubmitError();
        } else {
          this.$notify({
            title: 'Password changed successfully',
            type: 'is-success',
          });
          this.oldPassword = null;
          this.newPassword = null;
        }
      } catch (err) {
        this.handleSubmitError(err.message);
      } finally {
        this.isLoading = false;
      }
    },

    handleSubmitError(err) {
      const isIncorrentPasswordError = matchString(
        err,
        'authentication code mismatch',
      );

      this.$notify({
        title: isIncorrentPasswordError
          ? 'You entered incorrect password, try using a different one.'
          : 'Password was not changed.',
        type: 'is-danger',
      });
    },
  },
  mixins: [formMixin],
};
</script>
