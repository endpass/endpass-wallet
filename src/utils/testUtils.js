import { kebabCase } from 'lodash';
import { getFrom } from '@/utils/objects';

function convertDirectivesToAttrs(directives) {
  const attrs = {};

  if (directives) {
    directives.forEach(({ rawName, expression }) => {
      attrs[rawName] = expression;
    });
  }

  return attrs;
}

function convertListenersToAttrs(listeners) {
  const attrs = {};

  if (listeners) {
    Object.keys(listeners).forEach(key => {
      const handlers = [].concat(listeners[key]);
      const handler = handlers.find(h => {
        const name = getFrom(h, 'fns.name', 'name') || '';

        return name.includes('bound ');
      });

      if (handler) {
        attrs[`v-on:${key}`] = (handler.fns || handler).name;
      }
    });
  }

  return attrs;
}

function convertModelToAttrs(model) {
  return model ? { 'v-model': model.expression } : {};
}

function convertSlotToAttrs(slot) {
  return slot ? { slot } : {};
}

function generateElement(createElement, vnode) {
  return vnode.componentOptions
    ? generateElementFromComponent(createElement, vnode)
    : generateElementFromHTML(createElement, vnode);
}

function getChildrenElements(createElement, children) {
  return children
    ? children.map(child => generateElement(createElement, child))
    : undefined;
}

function generateElementFromComponent(
  createElement,
  { data, componentOptions },
) {
  const { attrs, directives, model } = data;
  const { tag, listeners, children } = componentOptions;
  const componentData = {
    ...data,
    attrs: {
      ...attrs,
      ...convertDirectivesToAttrs(directives),
      ...convertListenersToAttrs(listeners),
      ...convertModelToAttrs(model),
      class: data.staticClass,
    },
    hook: undefined,
    staticClass: undefined,
  };

  if (listeners) {
    componentData.on = Object.assign({}, listeners);
  }

  return createElement(
    tag,
    componentData,
    getChildrenElements(createElement, children),
  );
}

function generateElementFromHTML(createElement, vnode) {
  if (!vnode.data) {
    return vnode;
  }

  const { tag, data, children } = vnode;
  const { directives, on, model, attrs = {}, slot } = data;

  return createElement(
    tag,
    {
      ...data,
      attrs: {
        ...attrs,
        ...convertDirectivesToAttrs(directives),
        ...convertListenersToAttrs(on),
        ...convertModelToAttrs(model),
        ...convertSlotToAttrs(slot),
      },
    },
    getChildrenElements(createElement, children),
  );
}

export function generateStubs(Component) {
  return Object.values(Component.components).reduce((stubs, stubComponent) => {
    let elementName;

    if (stubComponent.name) {
      elementName = kebabCase(stubComponent.name);
    }

    return Object.assign(stubs, {
      [elementName]: {
        render(createElement) {
          return generateElementFromComponent(createElement, this.$vnode);
        },
      },
    });
  }, {});
}

export default {
  generateStubs,
};
