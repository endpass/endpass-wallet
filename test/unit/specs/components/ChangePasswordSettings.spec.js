import { shallow, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import VeeValidate from 'vee-validate';
import Notifications from 'vue-notification';

import ChangePasswordSettings from '@/components/ChangePasswordSettings';
import { generateStubs } from '@/utils/testUtils';

describe('ChangePasswordSettings', () => {
  let accountsActions;
  let storeOptions;
  let wrapper;

  beforeEach(() => {
    accountsActions = {
      updateWalletsWithNewPassword: jest.fn().mockResolvedValue(true),
      updateWallets: jest.fn(),
    };
    storeOptions = {
      modules: {
        accounts: {
          namespaced: true,
          actions: accountsActions,
          getters: {
            decryptedWallets: jest.fn(() => [{}]),
            hdWallet: jest.fn(() => ({})),
            encryptedHdWallet: jest.fn(() => ({})),
            encryptedWallets: jest.fn(() => [{}]),
          },
        },
      },
    };
    const localVue = createLocalVue();

    localVue.use(Vuex);
    localVue.use(Notifications);
    localVue.use(VeeValidate);

    const store = new Vuex.Store(storeOptions);

    wrapper = shallow(ChangePasswordSettings, {
      store,
      localVue,
      stubs: generateStubs(ChangePasswordSettings),
    });
  });

  describe('render', () => {
    it('should be a Vue component', () => {
      expect(wrapper.name()).toBe('ChangePasswordSettings');
      expect(wrapper.isVueInstance()).toBeTruthy();
    });

    it('should render the initial state of the component', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    describe('old password field', () => {
      it('should disable field', () => {
        wrapper.setData({
          isLoading: true,
        });

        expect(
          wrapper.find('v-password[name=oldPassword]').attributes().disabled,
        ).toBeTruthy();
      });

      it('should enable field', () => {
        wrapper.setData({
          isLoading: false,
        });

        expect(
          wrapper.find('v-password[name=oldPassword]').attributes().disabled,
        ).toBeUndefined();
      });
    });

    describe('new password field', () => {
      it('should disable field', () => {
        wrapper.setData({
          isLoading: true,
        });

        expect(
          wrapper.find('v-password[name=newPassword]').attributes().disabled,
        ).toBeTruthy();
      });

      it('should enable field', () => {
        wrapper.setData({
          isLoading: false,
        });

        expect(
          wrapper.find('v-password[name=newPassword]').attributes().disabled,
        ).toBeUndefined();
      });
    });

    describe('button', () => {
      it('should show loader', () => {
        wrapper.setData({
          isLoading: true,
        });

        expect(wrapper.find('v-button').attributes().loading).toBeTruthy();
      });

      it('should not show loader', () => {
        wrapper.setData({
          isLoading: false,
        });

        expect(wrapper.find('v-button').attributes().loading).toBeUndefined();
      });
    });
  });

  describe('methods', () => {
    describe('handleFormSubmit', () => {
      beforeEach(() => {
        jest.spyOn(wrapper.vm, '$notify');
        wrapper.setData({
          oldPassword: '1234',
          newPassword: '1234',
        });
      });

      it('should update all account wallets with new password if it is valid', async () => {
        expect.assertions(5);

        wrapper.vm.handleFormSubmit();
        await global.flushPromises();

        expect(accountsActions.updateWalletsWithNewPassword).toBeCalled();
        expect(wrapper.vm.$notify).toBeCalledWith({
          title: 'Password changed successfully',
          type: 'is-success',
        });
        expect(wrapper.vm.oldPassword).toBe(null);
        expect(wrapper.vm.newPassword).toBe(null);
        expect(wrapper.vm.isLoading).toBe(false);
      });

      it('should handle error if old password is incorrect and notify user about it', async () => {
        expect.assertions(5);

        const error = new Error('authentication code mismatch');

        accountsActions.updateWalletsWithNewPassword.mockRejectedValueOnce(
          error,
        );

        wrapper.vm.handleFormSubmit();
        await global.flushPromises();

        expect(accountsActions.updateWalletsWithNewPassword).toBeCalled();
        expect(wrapper.vm.$notify).toBeCalledWith({
          title: 'You entered incorrect password, try using a different one.',
          type: 'is-danger',
        });
        expect(wrapper.vm.oldPassword).toBe('1234');
        expect(wrapper.vm.newPassword).toBe('1234');
        expect(wrapper.vm.isLoading).toBe(false);
      });

      it('should handle error in other cases and notify user about it', async () => {
        expect.assertions(5);

        const error = new Error('foo');

        accountsActions.updateWalletsWithNewPassword.mockRejectedValueOnce(
          error,
        );

        wrapper.vm.handleFormSubmit();
        await global.flushPromises();

        expect(accountsActions.updateWalletsWithNewPassword).toBeCalled();
        expect(wrapper.vm.$notify).toBeCalledWith({
          title: 'Password was not changed.',
          type: 'is-danger',
        });
        expect(wrapper.vm.oldPassword).toBe('1234');
        expect(wrapper.vm.newPassword).toBe('1234');
        expect(wrapper.vm.isLoading).toBe(false);
      });

      it('should show error if changing was failed without errors', async () => {
        expect.assertions(5);

        accountsActions.updateWalletsWithNewPassword.mockResolvedValueOnce(
          false,
        );

        wrapper.vm.handleFormSubmit();
        await global.flushPromises();

        expect(accountsActions.updateWalletsWithNewPassword).toBeCalled();
        expect(wrapper.vm.$notify).toBeCalledWith({
          title: 'Password was not changed.',
          type: 'is-danger',
        });
        expect(wrapper.vm.oldPassword).toBe('1234');
        expect(wrapper.vm.newPassword).toBe('1234');
        expect(wrapper.vm.isLoading).toBe(false);
      });
    });
  });
});
