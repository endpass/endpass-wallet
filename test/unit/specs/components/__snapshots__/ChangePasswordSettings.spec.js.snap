// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangePasswordSettings render should render the initial state of the component 1`] = `
<v-form
  v-on:submit="bound handleFormSubmit"
>
  <label
    class="label"
  >
    Change Password
  </label>
   
  <v-password
    data-test="input-old-password"
    data-vv-as="Old Password"
    name="oldPassword"
    placeholder="Old Password"
    v-model="oldPassword"
    validator="required|min:8"
  />
   
  <v-password
    data-test="input-new-password"
    data-vv-as="New Password"
    name="newPassword"
    placeholder="New Password"
    v-model="newPassword"
    validator="required|min:8"
  />
   
  <v-button
    class-name="is-primary is-medium"
    data-test="submit-change-password"
  >
    
    Change Password
  
  </v-button>
</v-form>
`;
